### ISSUE TITLE
How to understand the argument "hard_negative_weight" ?

### ISSUE THREAD

**Issue by Doragd** [princeton-nlp/SimCSE#37]
In the hard negative part of `cos_sim`, the weight is added on the diagonal of the hard negative part matrix.
![image](https://user-images.githubusercontent.com/26213546/120495982-c4b92480-c3ef-11eb-9351-92951a55b8ee.png)

Is it used to add "complexity" to distinguishing hard examples?

**Comment by gaotianyu1350** [comment]
Hi,

Yes it is used for increasing the weight for hard negatives. But in our ablation study, we found that taking a larger weight for hard negatives does not bring significant difference.

**Comment by Doragd** [comment]
> Hi,
> 
> Yes it is used for increasing the weight for hard negatives. But in our ablation study, we found that taking a larger weight for hard negatives does not bring a significant difference.


Thank you for your answer. I'd like to ask a question that seems a little trivial.
why the `hard_negative_weight` is the logarithm of the weight? I think when the argument is input into the `nn.CrossEntropyLoss()`, which contains the `LogSoftmax` operation, it is really converted to a logarithm of itself.

**Comment by gaotianyu1350** [comment]
LogSoftmax first apply softmax (so the input is on the exponent) and then logarithm. You can see at [line 209](https://github.com/princeton-nlp/SimCSE/blob/fd6f5c6f0b2ad8009ffde57e030d3f41f60992be/simcse/models.py#L209) that the weight is added to the logits (inputs to softmax).

**Comment by Doragd** [comment]
Sorry to bother you again. The weight is added to the logits, which actually is input into the `nn.CrossEntropyLoss()`, which combines `LogSoftmax` and `NLLLoss` in one single class. That is to say, before inputting to `nn.CrossEntropyLoss()`, the weight is not the logarithm value. Is there a problem with my understanding?

**Comment by gaotianyu1350** [comment]
The first part is correct. I didn't get what's "the weight is not the logarithm value"

**Comment by Doragd** [comment]
> The first part is correct. I didn't get what's "the weight is not the logarithm value"

okay. Thanks for your careful and kind reply.
